### ISSUE TITLE
Pycloudlib does not cleanup on failed launch

### ISSUE THREAD

**Issue by hammerstefan** [canonical/pycloudlib#262]
Currently pycloudlib makes no attempt to clean up cloud resources on failure.  This can be confusing and financially dangerous for the end-user, if they are not aware they may need to manually clean up.  

Resources that might be orphaned:
* Running instances
* VPCs
* Resource Groups

In our recent integration with the pycloudlib OCI support, we ended up with roughly 20 orphaned instances per week.

**Current workaround**: 
Make sure to pass `wait=False` into every pycloudlib command, call `instance.wait()` directly, catch all exceptions, on exception attempt to `instance.delete()`

```py
cloud: pycloudlib.oci.cloud.OCI
try:
    instance = cloud.launch(wait=False)
    instance.wait()
except Exception:
    try:
        instance.delete(wait=False)
        instance.wait()
    except Exception:
        # ORPHANED INSTANCE MAYBE
        pass
```
This has a few problems:
    * Needing to call wait() separately even though we really just want start(wait=True) - we can't proceed until instance is running
    * Needing to wrap in two try/except blocks when none are desired
    * Have no idea if delete() failed because instance didn't exist to begin with, or if it failed to communicate with running instance and we still have an orphaned instance
    * `wait=True` is the default behavior, which means pycloudlib is orphaning tens of instances per week BY DEFAULT in our CD pipeline
    * pycloudlib is throwing poorly typed exceptions, requiring user to catch `Exception` class, which is always bad and fraught with errors (since you don't actually know WHAT caused an error)

Improvement options:

**Option 1**: Update documentation only

If this current behavior is intentional, then documentation and error messages should be improved to clearly reflect that.  The user must be aware during which steps resources could be orphaned.  When an error is raised, the error should indicate if resources are left behind, and those resource IDs should be reported.  Documentation should also contain how to use `wait=False` and manual `delete()` for automation use.

Pro: 
* Documentation improvements are generally good
* Least disruptive
Con:
* Still have unexpected, potentially costly default behavior
* Needing to add boilerplate wrappers to pycloudlib interaction 

**Option 2**: Automate cleanup on error, opt-in

Provide the user with a way to automate cleanup on lifecycle interaction error (start, delete, restart, shutdown). This could be runtime configuration to indicate pycloudlib should fully manage the instance lifecycle, and not leave orphaned instances.

Pro:
* Doesn't change existing default behavior
* Acceptable solution for pipelines where we never want orphans
* No need for `wait=False;wait;try;except;try;except` boilerplate around pycloudlib interactions
Con:
* Still have unexpected, potentially costly default behavior

**Option 3**: Automate cleanup on error, opt-out

Automate cleanup on lifecycle interaction error (start, delete, restart, shutdown)Default to automate. Never leave any orphaned resources unless explicitly asked to do so by the user. Could add a runtime configuration to indicate pycloudlib should leave instances running on failure.

Pro:
* Safe default behavior - don't surprise the user with unwanted orphaned instances running up billing
* No need for boilerplate
* Ideal solution for pipelines where we never want remaining instances
Con:
* Changing existing default behavior

My preference strongly lies with option-3, since that adheres to the principle of least surprise for the user.  I also think that part of option-1 is mandatory. Whatever solution we chose, documentation must be clear to the user of what to expect.  A user could be launching instances that cost > $1.00/hr. Either we help the user in cleaning up or we very clearly document how the user needs to clean up to avoid orphaned resources.

**Comment by aciba90** [comment]
I vote for either option-2 or option-3.

Actually, if there are concerns about breaking changes, we could implement 2, which does not break existing code, and switch the default at a later point to implement 3.

**Comment by TheRealFalcon** [comment]
One problem is that pycloudlib really can't know when it _should_ clean up instances. You mention lifecycle interaction errors (start, delete, restart, shutdown), but we can get exceptions on almost any API call (e.g., execute). Should every single API call take a `cleanup` or similar parameter in case something goes wrong? What determines when we should or shouldn't do automatic cleanup? Pycloudlib has a `delete()` method, so I don't think it's unreasonable for the caller to be responsible for deleting any instance resources created.

I think your example is a little more verbose than it needs to be. If you're wanting automatic cleanup via the API to look something like this:
```python
instance = cloud.launch(cleanup=True)
```

If you don't want exception handling code when something goes wrong, then the equivalent current way of doing it would be:
```python
instance = cloud.launch(wait=False)
try:
    instance.wait()
except Exception:
    instance.delete()  # wait=False shouldn't be needed unless you want it
```

If you do want exception handling code, both need to be a little more verbose. E.g.,
```python
try:
    instance = cloud.launch(cleanup=True)
except <LaunchException>:
    # Handle launch exception
except <DeleteException>:
    # Handle delete exception
```
vs
```python
try:
    instance = cloud.launch(wait=False)
    try:
        instance.wait()
    except Exception:
        instance.delete()
except <LaunchException>:
    # Handle launch exception
except <DeleteException>:
    # Handle delete exception
```

Is the addition of 3 lines of exception handling really that big of a problem? To me that is much preferable to littering the API with `cleanup` parameters. It's also much more consistent. "You're responsible for cleaning up your resources always" is a lot easier to remember than "You're responsible for cleaning up your resources except for these specific API calls".

One area I agree though, is that the default behavior of `cloud.launch()` makes it easy to unknowingly leak resources because you don't get that instance handle so a raised exception can't be properly handled. However, I would rather change the default to `wait=False` or perhaps remove the parameter entirely so it's possible to always cleanup the instance yourself.

In general, dealing with resources that need to be cleaned up is a perfect use case for a context manager in python (thanks @holmanb for the reminder). Perhaps a better compromise is the option to return a context manager on launch. Something like
```python
with cloud.launch() as instance:  # wait=True can still be the default here
    # do whatever with the instance
# Once we exit the `with` block, cleanup automatically occurs
```

**Comment by hammerstefan** [comment]
Sorry for late reply.

Regarding overall verbosity of different implementations, I care less about number of lines, and care more about the chance of accidentally leaking / orphaning cloud resources, especially with default parameters.

As you correctly pointed out, `cloud.launch()` is the primary culprit of potential orphaning since `wait=True` can throw an exception without the possibility of cleaning it up (since an instance will never be returned from `cloud.launch()` on `raise`). Once the user has an instance object, it is reasonable (and intuitive) to expect the user to clean it up themselves.

Regarding `cloud.launch()`, I'm good with either 
a) having it auto-clean if `wait==True`, or 
b) just removing the `wait` parameter (treat it like `wait=False` always) - I don't see value in keeping `wait` around if it doesn't auto-clean on, too high of a risk to orphan resources accidentally.

If you prefer no auto cleanup, then let's just remove `wait` from `cloud.launch()` and document clearly that users need to ensure they call `instance.delete()` for any instance launched, even if errors are raised. I see merit in the consistency of always requiring users handle cleanup.

I also see the point that adding `cleanup` argument to relevant `Instance` methods gets messy, and we can rely on user to clean up when they deem necessary.

**tl;dr**: Lets fix `cloud.launch()` to prevent accidental leaks
